From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 4 Jan 2020 11:00:00 +0100
Subject: [PATCH 2/2] libvips/pyvips#148


diff --git a/libvips/resample/reduceh.cpp b/libvips/resample/reduceh.cpp
index 1111111..2222222 100644
--- a/libvips/resample/reduceh.cpp
+++ b/libvips/resample/reduceh.cpp
@@ -486,7 +486,7 @@ vips_reduceh_build( VipsObject *object )
 	if( vips_embed( in, &t[1], 
 		reduceh->n_point / 2 - 1, 0, 
 		in->Xsize + reduceh->n_point - 1, in->Ysize,
-		"extend", VIPS_EXTEND_COPY,
+		"extend", VIPS_EXTEND_BLACK,
 		(void *) NULL ) )
 		return( -1 );
 	in = t[1];
diff --git a/libvips/resample/reducev.cpp b/libvips/resample/reducev.cpp
index 1111111..2222222 100644
--- a/libvips/resample/reducev.cpp
+++ b/libvips/resample/reducev.cpp
@@ -864,7 +864,7 @@ vips_reducev_build( VipsObject *object )
 	if( vips_embed( in, &t[1], 
 		0, reducev->n_point / 2 - 1, 
 		in->Xsize, in->Ysize + reducev->n_point - 1, 
-		"extend", VIPS_EXTEND_COPY,
+		"extend", VIPS_EXTEND_BLACK,
 		(void *) NULL ) )
 		return( -1 );
 	in = t[1];
